app/ide-desktop/lib/profiling/src/profiling.ts

```diff
@@ -0,0 +1,322 @@
+        let [log_level_name, interval_name] = [parts[0], parts[1]]
+        let profiler = parseLogLevel(log_level_name)
+        return [profiler, interval_name]
+    } catch (e) {
+        return undefined
+    }
+}
+
+// ======================
+// === IntervalHandle ===
+// ======================
+
+/**
+ * Handle that allows the setting of an `end` marker for a profiling interval that was started via
+ * the `start` function.
+ */
+class IntervalHandle {
+    interval_name: string
+    log_level: Profiler
+    released_status: ReleasedStatus
+
+    constructor(interval_name: string, log_level: Profiler) {
+        this.interval_name = interval_name
+        this.log_level = log_level
+        this.released_status = new ReleasedStatus()
+        this.released_status.released = false
+        handle_registry.register(
+            this,
+            new IntervalFinalizationRegistryEntry(interval_name, this.released_status)
+        )
+    }
+
+    /// Measure the interval.
+    end() {
+        this.log_level.end(this.interval_name)
+        this.released_status.released = true
+    }
+
+    /// Release the handle to manually call `end_interval` without emitting a warning.
+    release() {
+        this.released_status.released = true
+    }
+}
+
+/**
+ * Class that contains the information added to the `FinalizationRegistry` to decide whether to emit
+ * a warning when a `IntervalHandle` is garbage collected.
+ */
+class IntervalFinalizationRegistryEntry {
+    interval_name: string
+    release_status: ReleasedStatus
+
+    constructor(interval_name: string, release_status: ReleasedStatus) {
+        this.interval_name = interval_name
+        this.release_status = release_status
+    }
+}
+
+/** Class that contains information about the release status of an `IntervalHandle`. An
+ * `IntervalHandle` handle is released by calling `end` or `release`. If it is garbage collected
+ * without being released, we want to emit a warning. In order to be able to keep track of the
+ * release status of an `IntervalHandle` we need to keep the status boolean in a separate class. */
+class ReleasedStatus {
+    released: boolean
+
+    constructor() {
+        this.released = false
+    }
+}
+
+// We need to ts-ignore here as `FinalizationRegistry` is not recognised by TypeScript as being in
+// scope even though it is.
+// @ts-ignore
+const handle_registry = new FinalizationRegistry(heldValue => {
+    if (heldValue instanceof IntervalFinalizationRegistryEntry) {
+        if (!heldValue.release_status.released) {
+            console.warn(heldValue.interval_name + 'was dropped without a call to `measure`.')
+        }
+    }
+})
+
+// ===================
+// === Measurement ===
+// ===================
+
+/**
+ * Profiling measurement that contains the profiling information about a measured time interval.
+ */
+class Measurement {
+    readonly startTime: number
+    readonly duration: number
+    readonly logLevel: Profiler
+    readonly name: string
```

Comment:
can we also capture the file name and line number please? I'm almost sure TypeScript has utils for that.

Reply:
Done.